okhttp/src/main/java/okhttp3/internal/cache/CacheInterceptor.java

```diff
@@ -101,8 +101,23 @@ public CacheInterceptor(InternalCache cache) {
     // If we have a cache response too, then we're doing a conditional get.
     if (cacheResponse != null) {
       if (networkResponse.code() == HTTP_NOT_MODIFIED) {
+        Headers.Builder newResponseHeadersBuilder = combine(
+            cacheResponse.headers(),
+            networkResponse.headers()
+        ).newBuilder();
+
+        // If we have Content-Encoding IN the new response, but NOT in the old response our
+        // response body will be broken that's why we ignore the new Content-Encoding header. See
+        // https://github.com/square/okhttp/pull/3700 for more information.
+        String cachedContentEncoding = cacheResponse.headers().get("Content-Encoding");
+        if (cachedContentEncoding == null) {
+          newResponseHeadersBuilder.removeAll("Content-Encoding");
+        } else {
+          newResponseHeadersBuilder.set("Content-Encoding", cachedContentEncoding);
+        }
+
         Response response = cacheResponse.newBuilder()
-            .headers(combine(cacheResponse.headers(), networkResponse.headers()))
```

Comment:
Looking at this code more, combine is already single use and fairly specific to the use case of updating cached headers.  I suggest moving this logic into combine.

Reply:
@yschimke So like:

1. a list of ignored header fields as a parameter
2. ignoring those two headers by default within combine without extra parameters

what do you prefer?

Reply:
Ah I see "Content-Length" is handled similary at the moment.

Reply:
I moved "Content-Length" together with "Content-Type" and "Content-Encoding" into a new method isNotContentSpecificHeader and use this method in "combine" now to filter which header fields to ignore.